charts/argo-cd/templates/argocd-server/ingress-grpc.yaml

```
@@ -19,41 +15,66 @@ metadata:
   name: {{ template "argo-cd.server.fullname" . }}-grpc
   labels:
     {{- include "argo-cd.labels" (dict "context" . "component" .Values.server.name "name" .Values.server.name) | nindent 4 }}
-{{- if .Values.server.ingressGrpc.labels }}
-{{- toYaml .Values.server.ingressGrpc.labels | nindent 4 }}
-{{- end }}
+    {{- if .Values.server.ingressGrpc.labels }}
+    {{- toYaml .Values.server.ingressGrpc.labels | nindent 4 }}
+    {{- end }}
 spec:
+  {{- if eq (include "argo-cd.ingress.apiVersion" $) "networking.k8s.io/v1" }}
+  {{- with .Values.server.ingress.ingressClassName }}
+  ingressClassName: {{ . }}
+  {{- end }}
+  {{- end }}
   rules:
   {{- if .Values.server.ingressGrpc.hosts }}
-  {{- range $host := .Values.server.ingressGrpc.hosts }}
+    {{- range $host := .Values.server.ingressGrpc.hosts }}
     - host: {{ $host }}
       http:
         paths:
-  {{- if $extraPaths }}
-  {{- toYaml $extraPaths | nindent 10 }}
-  {{- end -}}
-  {{- range $p := $paths }}
+          {{- if $extraPaths }}
+          {{- toYaml $extraPaths | nindent 10 }}
+          {{- end -}}
+          {{- range $p := $paths }}
           - path: {{ $p }}
+            {{- if eq (include "argo-cd.ingress.apiVersion" $) "networking.k8s.io/v1" }}
+            pathType: Prefix
+            {{- end }}
             backend:
+              {{- if eq (include "argo-cd.ingress.apiVersion" $) "networking.k8s.io/v1" }}
+              service:
+                name: {{ $serviceName }}
+                port:
+                  number: {{ $servicePort }}
```

```suggestion
                {{- if or (kindIs "int" $servicePort) (kindIs "float64" $servicePort }}
                  number: {{ $servicePort }}
                {{- else }}
                  name: {{ $servicePort }}
                {{- end }}         
```
Note that we lose compatibility with this change, we need to evaluate if the port is a number or a port reference.

---

I agree but, do we also need to check for `float64`? What's the usecase?

---

Should be the question, why we check for `int`? By default these values are float64. And I thought we should also catch int.

---

@oliverbaehler There is never a case that you want to test for int when pulling from values.yaml. Example:
testchart/values.yaml
```yaml
tests:
  - 123
  - 123.123
  - "123"
```
templates/test.yaml
```
{{- range $i, $v := .Values.tests -}}
  {{- printf "# test %d KindOf: %s\n" $i (kindOf $v) -}}
{{- end -}}
```
output:
```
# Source: testchart/templates/test.yaml
# test 0 KindOf: float64
# test 1 KindOf: float64
# test 2 KindOf: string

---

@DirtyCajunRice Thanks for your input. I also tested it and couldn't really find a scenario where it becomes int. I implemented it that way in one of my libraries, it could be that it changes after multiple from/toyaml conversion, idk tbh :D. Anyway I adjusted my suggestions.

Do you have any creative idea how we could catch if a number is between quotes? As you showed in your example in quotes it will always be string and I am sure some people still have their number in quotes and that would be again breaking. I will also try to come up with something.

---

@oliverbaehler Yes you could always have templates that try to force conversions but IMO this is not the purpose of helm honestly. We version tag and notate breaking changes to ensure that proper usage, derived from informed updates, yield valid results. Realistically there is a bare minimum expectation of knowledge when someone is deploying applications via a templating interface (helm) derived from a templating language (go template/html + sprig) to an orchestration stack (k8s) over containerization (docker/containerd/etc).